algebra/euclidean_domain.lean

```diff
@@ -0,0 +1,316 @@
+/- 
+Copyright (c) 2018 Louis Carlin. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Louis Carlin
+
+Euclidean domains and Euclidean algorithm (extended to come)
+A lot is based on pre-existing code in mathlib for natural number gcds
+-/
+import tactic.ring
+
+universe u
+
+class euclidean_domain (α : Type u) extends integral_domain α :=
+(quotient : α → α → α)
+(remainder : α → α → α)
+(witness : ∀ a b, (quotient a b) * b + (remainder a b) = a) -- This could be changed to the same order as int.mod_add_div. We normally write qb+r rather than r + qb though.
+(valuation : α → ℕ)
+(val_mod : ∀ a b, b = 0 ∨  valuation (remainder a b) <  valuation b)
+(val_mul : ∀ a b, b = 0 ∨ valuation a ≤ valuation (a*b))
+(decidable_eq : ∀ a b : α, decidable (a=b))
+/-
+val_mul is often not a required in definitions of a euclidean domain since given the other properties we can show there is a (noncomputable) euclidean domain α with the property val_mul.
+So potentially this definition could be split into two different ones (euclidean_domain_weak and euclidean_domain_strong) with a noncomputable function from weak to strong.
+I've currently divided the lemmas into strong and weak depending on whether they require val_mul or not.
+-/
+
+namespace euclidean_domain
+variable {α : Type u}
+
+instance  euclidean_domain.has_decidable_eq [euclidean_domain α] (a b : α) : decidable (a = b) :=
+euclidean_domain.decidable_eq a b
+
+instance euclidean_domain_has_div [euclidean_domain α] : has_div α := 
+{ 
```

Comment:
remove the names for the instances. The command usually introduces its own names.

Also instead of
```lean
instance ... :=
{
  x := _ 
}
```

use the anonymous constructor (written using \< and \> in vscode and emacs.

```lean
instance ... := ⟨_⟩
```